staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,127 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageversion
+
+import (
+	apiserverinternalv1alpha1 "k8s.io/api/apiserverinternal/v1alpha1"
+	apiextensionshelpers "k8s.io/apiextensions-apiserver/pkg/apihelpers"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	genericstorageversion "k8s.io/apiserver/pkg/storageversion"
+)
+
+// Manager provides methods for updating StorageVersion for CRDs. It does
+// goroutine management to allow CRD storage version updates running in the
+// background and not blocking the caller.
+type Manager interface {
+	// EnqueueStorageVersionUpdate queues a StorageVesrion update for the given
+	// CRD and returns immediately. Optionally, the caller may specify a
+	// non-nil waitCh and/or a non-nil processedCh.
+	// A non-nil waitCh will block the StorageVersion update until waitCh is
+	// closed.
+	// The manager will close the non-nil processedCh if it finished
+	// processing the StorageVersion update (note that the update can either
+	// succeeded or failed).
+	UpdateStorageVersion(crd *apiextensionsv1.CustomResourceDefinition,
+		waitCh <-chan struct{}, processedCh chan<- struct{})
+}
+
+// manager implements the Manager interface.
+type manager struct {
+	// client is the client interface that manager uses to update
+	// StorageVersion objects.
+	client genericstorageversion.Client
+	// apiserverID is the ID of the apiserver that invokes this manager.
+	apiserverID string
+}
+
+// NewManager creates a CRD StorageVersion Manager.
+func NewManager(client genericstorageversion.Client, apiserverID string) Manager {
+	return &manager{
+		client:      client,
+		apiserverID: apiserverID,
+	}
+}
+
+// UpdateStorageVersion queues a StorageVesrion update for the given
+// CRD and returns immediately. Optionally, the caller may specify a
+// non-nil waitCh and/or a non-nil processedCh.
+// A non-nil waitCh will block the StorageVersion update until waitCh is
+// closed.
+// The manager will close the non-nil processedCh if it finished
+// processing the StorageVersion update (note that the update can either
+// succeeded or failed).
+func (m *manager) UpdateStorageVersion(crd *apiextensionsv1.CustomResourceDefinition,
+	waitCh <-chan struct{}, processedCh chan<- struct{}) {
+	// TODO: check if this means its waiting for waitCh to close
+	if waitCh != nil {
+		<-waitCh
```

Comment:
needs a select that wait on the ctx.Done to handle cancellation cases when wait isn't closed.  And when ctx.Done is closed, the processedCh should never close.

Reply:
Added a select block